service/qos/service_level_controller.cc

```diff
@@ -298,6 +298,22 @@ void service_level_controller::update_from_distributed_data(std::chrono::duratio
     }
 }
 
+future<> service_level_controller::stop_update_from_distributed_data() {
+    if (this_shard_id() != global_controller) {
+        return make_ready_future();
+    }
+    // abort the loop of the distributed data checking if it is running
+    if (!_global_controller_db->distributed_data_update.available()) {
+        if (!_global_controller_db->dist_data_update_aborter.abort_requested()) {
+            _global_controller_db->dist_data_update_aborter.request_abort();
+        }
+    }
+    _global_controller_db->distributed_data_update = _global_controller_db->distributed_data_update.get_future().then([this] () {
+        // "Recreate" abort source
```

Comment:
why?

Reply:
To be prepared to the scenario of `start->stop->start`
I guess you think it doesn't worth being prepared for.

> @eliransin
> 
> Did you see the problem even with the following patch? I think with the patch below the decommissioned node will be marked as down. Hopefully, the node will be excluded from your service level loop.
> 
> ```
> commit 0665d9c346170517e7e3fc16f45bcfac54e865cb
> Author: Asias He <asias@scylladb.com>
> Date:   Tue Jun 1 09:35:22 2021 +0800
> 
>     gossip: Handle nodes removed from live endpoints directly
>     
>     When a node is removed from the _live_endpoints list directly, e.g., a
>     node being decommissioned, it is possible the node might not be marked
>     as down in gossiper::failure_detector_loop_for_node loop before the loop
>     exits. When the gossiper::failure_detector_loop loop starts again, the
>     node will not be considered because it is not present in _live_endpoints
>     list any more. As a result, the node will not be marked as down though
>     gossiper::failure_detector_loop_for_node loop.
>     
>     To fix, we mark the nodes that are removed from _live_endpoints
>     lists as down in the gossiper::failure_detector_loop loop.
>     
>     Fixes #8712
>     
>     Closes #8770
> ```

@asias 
I only need to know on the current node so IIUC this will not affect my code.
However, what do you think about the rest of the code? I am considering to break it into two separate patches, one that
refactors the storage_proxy parallel_for_each usage (the first patch).
And all of the rest I am going to change to implement using your and @psarna pointers above. Do you think the refactor is
beneficial?